Userland/Libraries/LibWeb/Layout/BlockFormattingContext.cpp

```diff
@@ -361,35 +365,52 @@ void BlockFormattingContext::layout_block_level_box(Box const& box, BlockContain
         return;
 
     if (box.is_floating()) {
-        layout_floating_box(box, block_container, layout_mode, available_space);
+        layout_floating_box(box, block_container, layout_mode, available_space, m_margin_state.current_collapsed_margin_top() + current_y);
         bottom_of_lowest_margin_box = max(bottom_of_lowest_margin_box, box_state.offset.y() + box_state.content_height() + box_state.margin_box_bottom());
         return;
     }
 
     compute_width(box, available_space, layout_mode);
 
-    place_block_level_element_in_normal_flow_vertically(box);
-    place_block_level_element_in_normal_flow_horizontally(box, available_space);
-
     if (box_state.has_definite_height()) {
         compute_height(box, available_space);
     }
 
+    m_margin_state.add_margin(box_state.margin_top);
+    auto margin_top = m_margin_state.current_collapsed_margin_top();
+
+    place_block_level_element_in_normal_flow_vertically(box, current_y + box_state.border_box_top() + margin_top);
+    place_block_level_element_in_normal_flow_horizontally(box, available_space);
+
     OwnPtr<FormattingContext> independent_formatting_context;
     if (!box.is_replaced_box() && box.has_children()) {
         independent_formatting_context = create_independent_formatting_context_if_needed(m_state, box);
         if (independent_formatting_context) {
+            // Margins of elements that establish new formatting contexts do not collapse with their in-flow children
+            m_margin_state.reset();
+
             independent_formatting_context->run(box, layout_mode, box_state.available_inner_space_or_constraints_from(available_space));
         } else {
-            if (box.children_are_inline())
+            if (box.children_are_inline()) {
                 layout_inline_children(verify_cast<BlockContainer>(box), layout_mode, box_state.available_inner_space_or_constraints_from(available_space));
-            else
+            } else {
+                m_margin_state.reset();
                 layout_block_level_children(verify_cast<BlockContainer>(box), layout_mode, box_state.available_inner_space_or_constraints_from(available_space));
+            }
         }
     }
 
     compute_height(box, available_space);
 
+    // FIXME: it's too naive way to check if box margins collapse through
```

Comment:
If I tried to fix this, I wouldn't know what to do. Can we add some information about what the code *should* do, maybe even with a link to the relevant spec?

Reply:
makes sense, I updated this comment.